hashicorp/consul-k8s
# Support older versions of Consul in sync catalog

If namespaces are not enabled, use a Consul catalog API that is
available in older Consul versions instead of the API that's only
available in Consul 1.7+.

Design decisions:
* I've only implemented the adapter for the `NodeServiceList` endpoint rather than for the whole consul client API. This means `Syncer` needs both `ConsulClient` and `ConsulNodeServiceListClient` but it reduces the amount of code required vs needing to wrap the whole API of `ConsulClient`.
* Previously (before namespaces) the code used the `/catalog/services` endpoint to return just service names and then it filtered on those with the tag `k8s`. It never checked if those services were registered to the `k8s-sync` node. Rather than going back to that API, I've switched it to the `/catalog/node/:node_id` endpoint. I made this change because it matches what the namespaces code is doing by hitting the `/catalog/node-services/:node_id` endpoint. This makes the adapter easier to understand since it's adapting "list services for a node" instead of "list services for a node if namespaces, else list all services regardless of node". It is also backwards compatible because the only node the sync process ever registered services on is `k8s-sync`.
* Rather than having `ConsulClient` be a field in the adapting client, I'm making it part of the function signature. I made this decision because I liked that the Syncer was using the same underlying Consul client for all its calls. If it redefined its own `ConsulClient` field (it doesn't right now), then the clients would be out of sync if the adapting client had its own.
* I've changed the implementation of the namespaces-aware client to add a filter on tags. This filtering was originally being done in code but it seems better to use the proper API capabilities for this.

Manual Testing:
1. Consul OSS 1.6.3 - Use code before change, create a service manually with tag "k8s", see that it **does not** get deleted. This is because the catalog list call is failing and so we never populate the list of services to delete.
1. Consul OSS 1.6.3 - Use code after change, create a service manually with tag "k8s" on node k8s-sync, see that is deleted (proving that the bug is fixed)
1. Consul OSS 1.7.0 - Run with namespaces disabled, manually create a service with tag k8s on node k8s-sync, see that it is deleted.
1. Consul Enterprise 1.7.0 - Run with namespaces enabled syncing to one namespace. Change config to mirroring. See that services in old namespace are deleted.
